### ISSUE TITLE
Implement Happy Eyeballs to support IPv4/IPv6 combination

### ISSUE THREAD

**Issue by xsuchy** [shazow/urllib3#797]
In case host have IPv4 and IPv6 address but the route for one of them is broken, then user using urllib3 cannot connect to that URL.

This can be avoided (beside fixing the route, which is not always possible) by implementing Happy Eyeballs specification:
http://tools.ietf.org/html/rfc6555

**Comment by Lukasa** [comment]
Implementing Happy Eyeballs is really tricky for us, and potentially impossible, because urllib3 has a policy of not spawning its own threads. Because of the unpleasant way in which Python threads interact with the interpreter, libraries that spawn their own threads can have all kinds of unexpected effects on user code. This is a bad time, so we simply don't do it.

We could *in principle* use happy eyeballs temporarily switching the sockets to non-blocking mode, issuing all the connect requests at once, selecting on them all, and then switching them back to blocking once we know which one we want to use. That's worth considering, I suppose, but it'll be quite a chunk of work.

**Comment by Lukasa** [comment]
Ok, I actually believe we can do this.

The code is going to be crazy complex, and at this point we really will want to factor it out into a helper, but it's definitely do-able. I've written a version that isn't entirely spec-compliant, in part because it connects to every single DNS result at once (not entirely web-friendly), but it's do-able.

However, this is sufficiently complex that I'd want us to really factor this out aggressively. There's a lot of complexity involved in doing this properly: we'll want to cache successful results, we'll want to properly handle the case where the first two addresses we find both fail to connect, and we need to make sure that error reporting continues to function properly.

I think this is a good long term goal, but I also think it's one that is best suited to a BountySource style approach. I've uploaded my initial test implementation [here](https://github.com/Lukasa/urllib3/tree/happy-eyeballs): anyone wanting to take this work on is welcome to use it as a basis for it.

Alternatively, we can hope that I'll eventually get around to this!

**Comment by graingert** [comment]
ah I now see this is discussed here: https://github.com/urllib3/urllib3/pull/1000

**Comment by pquentin** [comment]
Removed the bounty per https://github.com/urllib3/urllib3/pull/3112#issuecomment-1702173568. We can revisit this issue after async support. #3038 is the best way to move this forward.

**Comment by pemensik** [comment]
Since #3038 is completed, is there possible way forward?

**Comment by pemensik** [comment]
It should be possible to do with just proper implementation of BSD sockets polling inside single thread. Just spawn IPv6 first, set a timeout delay for IPv4. Watch ipv6 socket for write readiness. If it does not happen soon, spawn also IPv4 socket. Both non-blocking, of course. Use ``poll(3)`` or similar to watch the first socket ready for writing. Close remaining sockets and proceed only with the ready socket normal way. But requires custom handling of timeouts and minimalistic event loop.

**Comment by pemensik** [comment]
That is true for Happy Eyeballs version 1. version 2 requires asynchronous DNS lookups, which are not provided by system API yet.

**Comment by pemensik** [comment]
It seems asyncio already has some support for asynchronous connections. Could it be used for socket connection, then changing the socket to blocking again and continuing with just connected socket? It would certainly complicate initial connection, but if the core is already implemented, it should not be so hard.

https://docs.python.org/3.12/library/asyncio-eventloop.html#opening-network-connections

But ideally user-friendy(er) interface should be possible directly over asyncio loops.
